aptos-move/move-examples/veiled_coin/sources/veiled_coin.move

```diff
@@ -0,0 +1,762 @@
+///
+/// Suppose Bob also registers and veils `b` of his own coins of type `T`.
+///
+/// Now Alice can use `fully_veiled_transfer` to send to Bob a secret amount `v` of coins from her veiled balance.
+/// This will, for the first time, properly hide both Alice's and Bob's veiled balance.
+/// The only information that an attacker (e.g., an Aptos validator) learns, is that Alice transferred an unknown amount
+/// `v` to Bob (including $v=0$), and as a result Alice's veiled balance is in a range [a-v, a] and Bob's veiled balance
+/// is in [b, b+v]`.
+///
+/// As more veiled transfers occur between more veiled accounts, the uncertainity on the balance of each account becomes
+/// larger and larger.
+///
+/// Lastly, users can easily withdraw veiled coins back into their public balance via `unveil`. Or, they can withdraw
+/// publicly into someone else's public balance via `unveil_to`.
+///
+/// ## Terminology
+///
+/// 1. *Veiled coin*: a coin whose value is secret; i.e., it is encrypted under the owner's public key.
+///
+/// 2. *Veiled amount*: any amount that is secret because it was encrypted under some public key.
+/// 3. *Committed amount*: any amount that is secret because it was committed to (rather thna encrypted).
+///
+/// 4. *Veiled transaction*: a transaction that hides its amount transferred; i.e., a transaction whose amount is veiled.
+///
+/// 5. *Veiled balance*: unlike a normal balance, a veiled balance is secret; i.e., it is encrypted under the account's
+///    public key.
+///
+/// 6. *ZKRP*: zero-knowledge range proofs; one of the key cryptographic ingredient in veiled coins which ensures users
+///    can withdraw secretely from their veiled balance without over-withdrawing.
+///
+/// ## Limitations
+///
+/// **WARNING:** This module is **experimental**! It is *NOT* production-ready. Specifically:
+///
+///  1. Deploying this module will likely lead to lost funds.
+///  2. This module has not been cryptographically-audited.
+///  3. The current implementation is vulnerable to _front-running attacks_ as described in the Zether paper [BAZB20].
+///  4. There is no integration with wallet software which, for veiled accounts, must maintain an additional ElGamal
+///    encryption keypair.
+///  5. There is no support for rotating the ElGamal encryption public key of a veiled account.
+///
+/// ## Veiled coin amounts as truncated `u32`'s
+///
+/// Veiled coin amounts must be specified as `u32`'s rather than `u64`'s as would be typical for normal coins in the
+/// Aptos framework. This is because coin amounts must be encrypted with an *efficient*, additively-homomorphic encryption
+/// scheme. Currently, our best candidate is ElGamal encryption in the exponent, which can only decrypt values around
+/// 32 bits or slightly larger.
+///
+/// Specifically, veiled coins are the middle 32 bits of the normal 64 bit coin values. In order to convert a `u32`
```

Comment:
Worth mentioning the number of bits past the decimal point for 32 bit values? About 10 bits to the right and 22 to the left.